File: algo-ds/CtCI-6th/chapter08/4_power_set.py
from typing import List

def power_set(arr: List[int]) -> List[List[int]]:
    rs = []

    for a in arr:
        temp = []
        for a2 in rs:
            print(f"a2 = {a2}")
            a3 = a2.append(a)
            temp.append(a3)
        rs.append([a])
        print(rs)
        if len(temp) > 0:
            rs.append(temp)

    return rs


if __name__ == "__main__":
    input_ = [1, 2, 3]
    rs = power_set(input_)
    print(rs)

File: algo-ds/CtCI-6th/chapter08/4_power_set.go
package main

import "fmt"


func main() {
    input_ := []int{1, 2, 3} 
    fmt.Println(powerSet(input_))
}

func powerSet(arr []int) [][]int {
    // Time and space: O(n*2^n)
    // 2^n subset and n is maximum number element of set
    rs := make([][]int, 0)

    for _, a1 := range arr {
        for _, a2 := range rs {
            a3 := append(a2, a1) 
            fmt.Println(a3)
            rs = append(rs, a3)
        }
        rs = append(rs, []int{a1})
    }
    return rs
}
